src/lib/operators/join_sort_merge.cpp

```diff
@@ -910,31 +891,73 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractJoinOperatorImpl {
+        for (const auto& row_id_right : *_null_rows_right) {
+          null_output_right->push_back(row_id_right);
+        }
+      }
+
+      if (!null_output_left->empty()) {
+        _output_pos_lists_left.push_back(null_output_left);
+        _output_pos_lists_right.push_back(null_output_right);
       }
     }
-    if (include_null_right) {
-      for (auto row_id_right : *_null_rows_right) {
-        output_left->push_back(NULL_ROW_ID);
-        output_right->push_back(row_id_right);
+
+    // Intermediate structure for output chunks (to avoid concurrent appending to table)
+    std::vector<std::shared_ptr<Chunk>> output_chunks(_output_pos_lists_left.size());
+
+    // Determine if writing output in parallel is necessary.
+    // As partitions ought to be roughly equally sized, looking at the first should be sufficient.
+    const auto write_output_concurrently = _cluster_count > 1 && _output_pos_lists_left[0]->size() > 10'000;
+
+    std::vector<std::shared_ptr<AbstractTask>> output_jobs;
+    output_jobs.reserve(_output_pos_lists_left.size());
+    for (auto pos_list_id = size_t{0}; pos_list_id < _output_pos_lists_left.size(); ++pos_list_id) {
+      if (_output_pos_lists_left[pos_list_id]->empty() && _output_pos_lists_right[pos_list_id]->empty()) {
+        continue;
+      }
+
+      auto write_output_chunk = [this, pos_list_id, &output_chunks] {
+        Segments segments;
+        _add_output_segments(segments, _sort_merge_join.input_table_left(), _output_pos_lists_left[pos_list_id]);
+        _add_output_segments(segments, _sort_merge_join.input_table_right(), _output_pos_lists_right[pos_list_id]);
+        output_chunks[pos_list_id] = std::make_shared<Chunk>(std::move(segments));
+      };
+
+      if (write_output_concurrently) {
+        auto job = std::make_shared<JobTask>(write_output_chunk);
+        output_jobs.push_back(job);
+        output_jobs.back()->schedule();
+      } else {
+        write_output_chunk();
       }
     }
 
-    // Add the segments from both input tables to the output
-    Segments output_segments;
-    _add_output_segments(output_segments, _sort_merge_join.input_table_left(), output_left);
-    _add_output_segments(output_segments, _sort_merge_join.input_table_right(), output_right);
+    // Wait for all chunk creation tasks to finish
+    Hyrise::get().scheduler()->wait_for_tasks(output_jobs);
+
+    // Remove empty chunks which occur due to empty radix clusters or not matching tuples of clusters.
+    output_chunks.erase(
+        std::remove_if(output_chunks.begin(), output_chunks.end(),
+                       [](const auto& output_chunk) { return !output_chunk || output_chunk->size() == 0; }),
+        output_chunks.end());
 
-    // Build the output_table with one Chunk
-    return _sort_merge_join._build_output_table({std::make_shared<Chunk>(output_segments)});
+    // TODO(anyone): when setting chunk ordered by flags, check if output is really sorted for NULL-adding joins.
```

Comment:
Are you suggesting adding a sanity check here? Is this something that is potentially broken?

Reply:
It is a hint for anyone (i.e., the exploitation seminar group) to not blindly set the flag for all chunks.
I am not certain, that such chunks are sorted. So “anybody” should either not set flags for NULLable  segments or take a deeper look (maybe with a DebugAssert on is_sorted) into nulling joins.

Reply:
Is this really an actionable TODO then or is it more of a warning?

Reply:
It’s a to do, but I’ll talk to my seminar group. Probably van be removed as they are aware of it anyways.

Reply:
Pinging @j-tr, @MasterCarl, @Boffmann here. I will remove the comment as it is not obvious what I am talking about.
Still, I want you to be aware of the "NULL chunks" for your own work on setting the sort flags.